src/devtools/client/webconsole/actions/messages.ts

```diff
@@ -231,3 +239,128 @@ export const messagesClearEvaluations = messageEvaluationsCleared;
+    const lastFetchedForFocusRegion = getLastFetchedForFocusRegion(state);
+
+    // Soft Focus: The frontend only needs to refetch data if:
+    // 1. The most recent time it requested data "overflowed" (too many messages to send them all), or
+    // 2. The new focus region is outside of the most recent region we fetched messages for.
+    //
+    // There are two things to note about the second bullet point above:
+    // 1. When devtools is first opened, there is no focused region.
+    //    This is equivalent to focusing on the entire timeline, so we often won't need to refetch messages when focusing for the first time.
+    // 2. We shouldn't compare the new focus region to the most recent focus region,
+    //    but rather to the most recent focus region that we fetched messages for (the entire timeline in many cases).
+    //    If we don't need to refetch after zooming in, then we won't need to refetch after zooming back out either,
+    //    (unless our fetches have overflowed at some point).
+    if (!didOverflow) {
+      if (isNewFocusRegionSubset(lastFetchedForFocusRegion, focusRegion)) {
+        // We already have all of the console logs for the new region.
+        // We can skip running a new analysis.
+        return;
+      }
+    }
+
+    dispatch(clearMessages());
+
+    const sessionEndpoint = await sendMessage("Session.getEndpoint", {}, ThreadFront.sessionId!);
+
+    const startTime = focusRegion ? focusRegion.startTime : 0;
+    const endTime = focusRegion ? focusRegion.endTime : sessionEndpoint.endpoint.time;
+
+    let beginPoint: string | null = null;
+    let endPoint: string | null = null;
+
+    if (sessionEndpoint.endpoint.time === endTime) {
+      const pointNearBeginning = await sendMessage(
+        "Session.getPointNearTime",
+        {
+          time: startTime,
+        },
+        ThreadFront.sessionId!
+      );
+
+      beginPoint = pointNearBeginning.point.point;
+      endPoint = sessionEndpoint.endpoint.point;
+    } else {
+      const [pointNearBeginning, pointNearEnd] = await Promise.all([
+        sendMessage(
+          "Session.getPointNearTime",
+          {
+            time: startTime,
+          },
+          ThreadFront.sessionId!
+        ),
+        sendMessage("Session.getPointNearTime", { time: endTime }, ThreadFront.sessionId!),
+      ]);
+
+      beginPoint = pointNearBeginning.point.point;
+      endPoint = pointNearEnd.point.point;
+    }
+
+    const { messages, overflow } = await sendMessage(
+      "Console.findMessagesInRange",
+      {
+        range: { begin: beginPoint, end: endPoint },
+      },
+      ThreadFront.sessionId!
+    );
+
+    // Store the result of the new analysis window for "soft focus" comparison next time.
+    dispatch(setLastFetchedForFocusRegion(focusRegion));
+    dispatch(setConsoleOverflowed(overflow === true));
+
+    // Copied from ThreadFront.findConsoleMessages():
+    // TODO [bvaughn] Would be nice if this shared code with ThreadFront.
```

Comment:
:question: is the issue that messages _normally_ come over the websocket via the standard "get some messages" request, and here we have an alternate "fetch messages" request that isn't used anywhere else, so we have to duplicate the processing?

Reply:
Yup! And if we don't do the mapping of message values to `ValueFront`s, it can cause things to break later (see https://github.com/RecordReplay/devtools/pull/6556#issuecomment-1134621051)